### DISCUSSION TITLE
How do I get Immunizations for a patient?

### DISCUSSION THREAD

**Discussion by johnkwaters** [FirelyTeam/firely-net-sdk#1967] [Q&A] [answered]
Looking at the network tab in a Fire product, I see

https://fhir-st-mpi-clinical.staging.healthconcourse.com/Immunization?patient=HC-000-1569

How do I issue a search for Immunizations through Firely? 

This is not allowed:

var immunizations = await client.SearchAsync<Immunization>(new SearchParams("patient", patientId));

**Comment by marcovisserFurore** [comment]
Hi @johnkwaters,

Thanks for question.

You have to tell the client which type of resource you are searching. So it would be:

```c#
   var immunizations = await client.SearchAsync<Immunization>(new SearchParams("patient", patientId));
```

**Comment by johnkwaters** [comment]
But I am not allowed to:

![image](https://user-images.githubusercontent.com/3422637/153959309-64ce9da7-f641-4443-9997-958cd51b3d29.png)

**Reply by marcovisserFurore** [reply]
Hi @johnkwaters,

I see that you are using another `Immunization` class (`FHIR2HC.Immunization`), which is not an official SDK class. So this should work:

```c#
var immunizations = await client.SearchAsync<Hl7.Fhir.Model.Immunization>(new SearchParams("patient", patientId));
```
And when you want to iterate only the `Immunization` resources (after you searched for them), you can do this:

```c#
  foreach(var resource in immunizations.Entry.ByResourceType<Hl7.Fhir.Model.Immunization>())
  {
      // process the resource
  }
```

Happy coding!

**Comment by johnkwaters** [comment, marked as answer]
Ah - good catch, and thanks for the ByResourceType<> tip

**Comment by johnkwaters** [comment]
Another question - not related, sorry! - when I get observations (vital signs) I see encounter is null for all of them. Yet some of them do seem to be related to one Encounter I get when fetching Encounters. How do I query so I get the Observations with an Encounter, or the Encounter with Observations?

**Reply by marcovisserFurore** [reply]
Perhaps next time, you can start a new question, so we don't mix up the topics.

Getting `Observations` that have a subject:
```c#
var result = client.Search<Observation>(new[] { "subject:missing=false" });
```
Getting `Observations` and their corresponding subjects in one bundle:
```c#
 var result = client.Search<Observation>(includes: new[] { "Observation:subject" });
 ```

See [here ](http://hl7.org/fhir/search.html) or [here](https://docs.fire.ly/projects/Firely-NET-SDK/client/search.html) for more information about search.

**Reply by johnkwaters** [reply]
Thanks. That first link doesn't work BTW.

**Reply by johnkwaters** [reply]
I tried 

    var sp = new SearchParams("subject", patientId)
        .Include("Observation:subject");
and

    var sp = new SearchParams("subject", patientId)
        .Include("Observation:encounter");

then

    var observations = await client.SearchAsync<Observation>(sp);
    
but neither returned an Encounter with the Observations

And is there an inverse - to get the Encounter and include Observations?

**Reply by mbaltus** [reply]
Corrected the link in Marco's reply.

@johnkwaters If the Observation.encounter fields for your Observations are all null as you've mentioned in your question, you will not get Encounters in the result bundle when you use the Include, because there are none linked to from Observation.
However, you do mention "some of them do seem to be related to one Encounter" when you look at the Encounters. Encounters can only link to Observations through the reasonReference field, and the corresponding search parameter is 'reason-reference'.  So, searching for Observations, you should be able to do a reverse include like this:

``` csharp
var sp = new SearchParams("subject", patientId)
    .ReverseInclude("Encounter:reason-reference");
```

If you do the search on Encounters, it will be the same search parameter, but a regular include:

``` csharp
var sp = new SearchParams().Include("Encounter:reason-reference");
```
